docs/typescript.md

```diff
@@ -1,14 +1,209 @@
-# TypeScript Guidelines
+# Typescript Guidelines
 
-These guidelines specifically apply to TypeScript.
+**_Purpose_**: Establish best practices for writing robust, maintainable, and efficient TypeScript code.
 
-## Provide explicit return types for functions/methods
+**_Audience_**: Internal and external developers using TypeScript in Metamask.
+
+## **Type**
+
+By defining expected object structures with interfaces, TypeScript promotes developer intent and unlocks powerful IDE features like accurate completion and context-aware navigation.
```

Comment:
(continued from https://github.com/MetaMask/contributor-docs/pull/80#discussion_r1506360915)

On second glance, we do use interfaces in the sense that you're describing it. And I think establishing contracts for classes is a development pattern worth encouraging. We just consistently use type alias syntax instead of interface syntax to write all types, including class "interfaces." Sorry for the confusion here. 

How about we split this into two sections?

```suggestion
## `implements` keyword

The `implements` keyword enables us to define and enforce interfaces, i.e. strict contracts consisting of expected object and class properties and abstract method signatures. 

Writing an interface to establish the specifications of a class that external code can interact with without being aware of internal implementation details is encouraged as sound OOP development practice. 

Here's an abbreviated example from `@metamask/polling-controller` of an interface being used to define one of our most important constructs.

\```ts
export type IPollingController = {
  ...
}
...
export function AbstractPollingControllerBaseMixin<TBase extends Constructor>(
  Base: TBase,
) {
  abstract class AbstractPollingControllerBase
    extends Base
    implements IPollingController
  { ... }
  return AbstractPollingControllerBase  
}
\```

The concept of the interface as discussed in this section is not to be confused with interface syntax as opposed to type alias. Note that in the above example, the `IPollingController` interface is defined as a type alias, not using the `interface` keyword. 

## Always prefer Type Aliases over Interface syntax

We enforce consistent and exclusive usage of type aliases over interface syntax for several reasons:

- The capabilities of type aliases is a strict superset of those of interfaces. 
  - Crucially, `extends`, `implements` are also supported by type aliases.
  - Declaration merging is the only exception, but we have no use case for this feature that cannot be substituted by using type intersections.
- Unlike interfaces, type aliases extend `Record` and have an index signature of `string` by default, which makes them compatible with our Json-serializable types (most notably `Record<string, Json>`). 
- Interfaces do not support multiple inheritance, while type aliases can be freely merged using the intersection (`&`) operator.
```

(remove the backticks around the fenced code block)

This is intended as an initial draft. Feel free to edit.

Reply:
Removing the second entry is also acceptable since `@typescript-eslint/consistent-type-definitions` does a good job of enforcing the practice without everyone having to know the reasoning behind its adoption. 

In general, the guidelines doesn't need to cover conventions that we already enforce using linter rules or static code analysis in too much detail. This doc isn't intended as a comprehensive collection of every preference we have agreed upon. The priority should be on covering principles or processes that aren't reducible to simple rulesets and rely on a deeper understanding from the developers to be effective.

Reply:
Thanks for all the recommendations, i've added your suggestions, and changed the section title as `Implements`